File: advent-of-code/R/2016_20.R
library(tidyverse)



# 

df <-
  tibble(line = read_lines(input))  %>%
  transmute(
    start = str_extract(line, "^\\d+") %>% parse_number(),
    end = str_extract(line, "\\d+$") %>% parse_number()
  )
df

find_min <- function() {
  i <- 0
  repeat {
    new_end <-
      df %>%
      filter(start <= i, end >= i) %>%
      pull(end) %>%
      max()
    
    if (!is.finite(new_end)) {
      return(i)
    }
    i <- new_end + 1
  }
}

find_min()

count_allowed <- function(max_allowed = 4294967295) {
  allowed <- 0
  i <- 0
  while(i <= max_allowed) {
    new_end <-
      df %>%
      filter(start <= i, end >= i) %>%
      pull(end) %>%
      max()
    
    if (!is.finite(new_end)) {
      new_end <-
        df %>%
        filter(start > i) %>%
        pull(start) %>%
        min()
      
      if (!is.finite(new_end)) {
        return (allowed + max_allowed - i + 1)
      }
      
      allowed <- allowed + (new_end - i)
      i <- new_end
    } else {
      i <- new_end + 1
    }
  }
  allowed
}

count_allowed()

File: advent-of-code/python/2016_20.py
from aocd import get_data

inp = get_data(day=20, year=2016)

def solve(ends, max_allowed = 4294967295):
  i = 0
  count_allowed = 0
  first_allowed = None
  
  while i <= max_allowed:
    new_end = max(end for start, end in ends if start <= i)
    if new_end <= i:
      first_allowed = first_allowed or i
      new_end = min((start for start, end in ends if start > i), default=max_allowed)
      count_allowed += new_end - i
    i = new_end + 1
  
  return first_allowed, count_allowed

first_allowed, count_allowed = solve([[int(x) for x in line.split('-')] for line in inp.split('\n')])

answer = first_allowed
answer

answer = count_allowed
answer
